### PR TITLE
Register a different context at runtime for a specific event

### PR DESCRIPTION
**Pull request by spinettaro**
added mechanism to register a new context
added relative test

### CODE REVIEW

**Comment by dkounal** [comment]
I think that this is much more flexible. You can have different instance contexts (more than one per subscriber) and per interface-event.

File: source/EventBus.Core.pas [hunk 1]
```diff
@@ -379,6 +444,33 @@ procedure TEventBus.Subscribe<T>(ASubscriber: TObject; ASubscriberMethod: TSubsc
...
+  if not LSubscriberToMethodCategoriesMap.TryGetValue(ASubscriber, LMethodCategories) then
+    exit(False);
+
+  LQualifiedEventName:= TInterfaceHelper.GetQualifiedName( AEvent);
+  for LMethodCategory in LMethodCategories do
+  begin
+    if LMethodCategory.Contains( LQualifiedEventName) then
+    begin
+        AMethodCategory:= LMethodCategory;
+        Exit(true);
```
Comment (wxinix) [inline source/EventBus.Core.pas]: Exit(True);

File: source/EventBus.Core.pas [hunk 2]
```diff
@@ -379,6 +444,33 @@ procedure TEventBus.Subscribe<T>(ASubscriber: TObject; ASubscriberMethod: TSubsc
...
+  LMethodCategory: String;
+  LQualifiedEventName: String;
+begin
+  if not FSubscriberToCategoriesByAttrName.TryGetValue(SubscribeAttribute.ClassName, LSubscriberToMethodCategoriesMap) then
+    exit(False);
+
+  if not LSubscriberToMethodCategoriesMap.TryGetValue(ASubscriber, LMethodCategories) then
+    exit(False);
+
+  LQualifiedEventName:= TInterfaceHelper.GetQualifiedName( AEvent);
```
Comment (wxinix) [inline source/EventBus.Core.pas]: GetQualifiedName(AEvent);

File: source/EventBus.Core.pas [hunk 3]
```diff
@@ -379,6 +444,33 @@ procedure TEventBus.Subscribe<T>(ASubscriber: TObject; ASubscriberMethod: TSubsc
   LCategories.Add(LCategory);
 end;
 
+function TEventBus.TryGetMethodCategory<T>(ASubscriber: TObject; AEvent: IInterface; out AMethodCategory: string): Boolean;
+var
+  LSubscriberToMethodCategoriesMap: TSubscriberToMethodCategoriesMap;
+  LMethodCategories: TMethodCategories;
+  LMethodCategory: String;
```
Comment (wxinix) [inline source/EventBus.Core.pas]: LMethodCategory: string

File: source/EventBus.Core.pas [hunk 4]
```diff
@@ -379,6 +444,33 @@ procedure TEventBus.Subscribe<T>(ASubscriber: TObject; ASubscriberMethod: TSubsc
   LCategories.Add(LCategory);
 end;
 
+function TEventBus.TryGetMethodCategory<T>(ASubscriber: TObject; AEvent: IInterface; out AMethodCategory: string): Boolean;
+var
+  LSubscriberToMethodCategoriesMap: TSubscriberToMethodCategoriesMap;
+  LMethodCategories: TMethodCategories;
+  LMethodCategory: String;
+  LQualifiedEventName: String;
```
Comment (wxinix) [inline source/EventBus.Core.pas]: LQualifiedEventName: string

**Review by wxinix — COMMENTED** [review]
This is the most flexible way.   Some minor consistency issue:  Nil -> nil,  String-> string, exit -> Exit, true -> True

My suggestion (just a suggestion) - for  IEventbus.RegisterNewContext, I think it might be better if we make it a method of a separate interface ISubscriberRegisteration, which should be returned by IEventBus.RegisterSubscriberForEvent

```
ISubscriberRegistration = interface
   AddContext(AEventType, AContext): ISubscriberRegistration;
end;
```

Then we can have some fluent-style of call:
RegisterSubscriberForEvent(obj)
.AddContext
.AddContext
.AddContext....

**Comment by wxinix** [comment]
also just a perspective  - if the user specifies the compile time attribute explicitly, that should be treated as a "contract" 
so the runtime context can be specified only for those methods that are "allowed" to be override by runtime context - 

to that end,  we need the TContextOption = (UseAttributeContext, UseInstanceContext)

Only those methods marked with UseInstanceContext can be assigned a runtime context?  If they don't have runtime context specified, then they take the default (global) context , by an empty string.

But this is just a perspective.

**Comment by wxinix** [comment]
> Then we can have some fluent-style of call:
> RegisterSubscriberForEvent(obj)
> .AddContext
> .AddContext
> .AddContext....

This design would encourage the user to add new context at that SAME call site of RegisterSubscriberForEvent

**Comment by dkounal** [comment]
The functionality is my major concern, but I do not say No to anything that will make it more user friendly.
I think that as it is now, the TContextOption is not needed. But this is just an opinion.

**Comment by wxinix** [comment]
That is OK. No problem.

**Comment by spinettaro** [comment]
@wxinix  I agree with @dkounal . When you call **RegisterNewContext** you deliberately choose to change the context. I think the additional checks are a kind of "over-checking" .  I like the fluent interface idea :)

**Comment by wxinix** [comment]
`TContextOption` is probably not needed. I agree. 

For the extra interface` ISubscriberRegistration`, which is for facilitating fluent chained calls, I suggest it as follows (if this idea of fluent style is accepted):

```
ISubscriberRegistration = interface
   procedure UseContext(const ANewContext: string; AEventType): ISubscriberRegistration; 
   procedure UseContextForAll(const ANewContext: string): ISubscriberRegistration; 
end;

```
The interface method `UseContextForAll` doesn’t have the second parameter for a specific EventType —— hence it will be applied to **all** subscriber methods.

Also I`SubscriberRegistration` would best be a private nested type in `IEventBus`, limiting its scope to the **_call site_** of `IEventBus.RegisterSubscriberForEvent` only. @spinettaro

**Pull request merged**